File: CS50/Problems/sentimental-credit/credit.py
# import functions
from cs50 import get_int

# card num
card = get_int("Card Number: ")

# vars
d = 0
count = 0
sum1 = 0
sum2 = 0
chsm = 0
switch = True
d13 = 0
d14 = 0
d15 = 0
d16 = 0

# checksum
while card > 0:
    if card >= 10:
        d = int(card % 10)
        card = int(card / 10)
        count += 1
    else:
        d = int(card)
        card = int(card / 10)
        count += 1
    
    # 1st 2nd switch    
    if switch == True:
        sum1 = sum1 + d
        switch = False
    elif switch == False:
        d1 = d * 2
        if d1 >= 10:
            d2 = int(d1 % 10)
            d1 = int(d1 / 10)
            sum2 = sum2 + d1 + d2
        else:
            sum2 = sum2 + d1
        switch = True
    
    # digits for card type
    if count == 13:
        d13 = d
    if count == 14:
        d14 = d
    if count == 15:
        d15 = d
    if count == 16:
        d16 = d

# checksum result
chsm = sum1 + sum2
chsm = chsm % 10

# display cardtype
if chsm == 0 and count >= 13:
    if count == 16:
        # master check
        if d16 == 5 and d15 <= 5:
            print("MASTERCARD")
        # visa check
        elif d16 == 4:
            print("VISA")
        else:
            print("INVALID")
    # AMEX check
    elif count == 15:
        if d15 == 3:
            if d14 == 4 or d14 == 7:
                print("AMEX")
            else: 
                print("INVALID")
    # visa check
    elif count == 13:
        if d13 == 4:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")

File: CS50/Problems/credit/credit.c
# include <cs50.h>
# include <stdio.h>

int main(void)
{
    // initializing variables
    int d = 0;
    int d1 = 0;
    int sum1 = 0;
    int sum2 = 0;
    int chsm = 0;
    int switch1 = 0;
    int digit_counter = 0;
    int d13 = 0;
    int d14 = 0;
    int d15 = 0;
    int d16 = 0;

    //get card number
    long card = get_long("Card Number: ");

    // checksum
    // get digits
    while (card > 0)
    {
        if (card >= 10)
        {
            d = card % 10;
            card /= 10;
            digit_counter++;
        }
        else
        {
            d = card;
            card /= 10;
            digit_counter++;
            
        }

        // every 2nd digit
        switch (switch1)
        {
            case 0:
                sum1 = sum1 + d;
                switch1++;
                break;

            case 1:
                d1 = d * 2;
                if (d1 >= 10)
                {
                    int d2 = d1 % 10;
                    d1 /= 10;
                    sum2 = sum2 + d1 + d2;
                }
                else
                {
                    sum2 = sum2 + d1;
                }
                switch1--;
                break;
        }

        // digit counter for cardtype
        if (digit_counter == 13)
        {
            d13 = d;
        }
        if (digit_counter == 14)
        {
            d14 = d;
        }
        if (digit_counter == 15)
        {
            d15 = d;
        }
        if (digit_counter == 16)
        {
            d16 = d;
        }
    }

    // checksum last digit
    chsm = sum1 + sum2;
    chsm = chsm % 10;

    // cardtype
    if (chsm == 0 && digit_counter >= 13)
    {
        // check lenght
        if (digit_counter == 16)
        {
            // display master
            if (d16 == 5 && d15 <= 5)
            {
                printf("MASTERCARD\n");
            }
            // visa
            else if (d16 == 4)
            {
                printf("VISA\n");
            }
            //invalid
            else
            {
                printf("INVALID\n");
            }
        }
        // check lenght
        else if (digit_counter == 15)
        {
            if (d15 == 3)
            {
                //display names amex / invalid
                if (d14 == 4 || d14 == 7)
                {
                    printf("AMEX\n");
                    
                }
                else
                {
                    printf("INVALID\n");
                    
                }
            }
        }
        //check lenght
        else if (digit_counter == 13)
        {
            //display visa / invalid
            if (d13 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    // display invalid
    else
    {
        printf("INVALID\n");
    }
}